core/amber/src/main/python/core/runnables/main_loop.py

```diff
@@ -152,6 +160,11 @@ def receive(self, next_entry: QueueElement) -> None:
             self._process_channel_marker_payload,
         )
 
+        # Upon receiving EndWorker control message, terminate this worker.
+        if self.context.state_manager.get_current_state() == WorkerState.TERMINATED:
+            logger.info(f"Python worker {self.context.worker_id} exited.")
+            self.worker_shutdown()
```

Comment:
it turned to TERMINATED state before shutdown() is invoked. Maybe better to change the state to `TERMINATING`?